packages/agent/src/oidc.ts

```diff
@@ -0,0 +1,703 @@
+  code_challenge: string;
+}) {
+  const [
+    protectedHeaderB64U,
+    ,
+    nonceB64U,
+    ciphertextB64U,
+    authenticationTagB64U,
+  ] = jwe.split('.');
+
+  const codeChallengeBytes = Convert.base64Url(code_challenge).toUint8Array();
+  const protectedHeader = Convert.base64Url(protectedHeaderB64U).toUint8Array();
+  const additionalData = protectedHeader;
+  const nonce = Convert.base64Url(nonceB64U).toUint8Array();
+  const ciphertext = Convert.base64Url(ciphertextB64U).toUint8Array();
+  const authenticationTag = Convert.base64Url(
+    authenticationTagB64U
+  ).toUint8Array();
+
+  // The cipher expects the encrypted data and tag to be concatenated.
+  const ciphertextAndTag = new Uint8Array([
+    ...ciphertext,
+    ...authenticationTag,
+  ]);
+  const chacha = xchacha20poly1305(codeChallengeBytes, nonce, additionalData);
+  const decryptedJwtBytes = chacha.decrypt(ciphertextAndTag);
+  const jwt = Convert.uint8Array(decryptedJwtBytes).toString();
+
+  return jwt;
+}
+
+/**
+ * Used by the client to decrypt the jwe obtained from the auth server which contains
+ * the {@link Web5ConnectAuthResponse} that was sent by the provider to the auth server.
+ *
+ * @async
+ * @param {BearerDid} clientDid - The did that was initially used by the client for ECDH at connect init.
+ * @param {string} jwe - The encrypted data as a jwe.
+ * @param {string} pin - The pin that was obtained from the user.
+ */
+async function decryptAuthResponse(
+  clientDid: BearerDid,
+  jwe: string,
+  pin: string
+) {
+  const [
+    protectedHeaderB64U,
+    ,
+    nonceB64U,
+    ciphertextB64U,
+    authenticationTagB64U,
+  ] = jwe.split('.');
+
+  // get the Provider's public key from the header
+  const header = Convert.base64Url(protectedHeaderB64U).toObject() as Jwk;
+  const providerDid = await DidDht.resolve(header.kid!.split('#')[0]);
+
+  // derive ECDH shared key using the provider's public key and our clientDid private key
+  const sharedKey = await deriveSharedKey(clientDid, providerDid.didDocument!);
+
+  // add the pin to the AAD
+  const additionalData = { ...header, pin: pin };
+  const AAD = Convert.object(additionalData).toUint8Array();
+
+  const nonce = Convert.base64Url(nonceB64U).toUint8Array();
+  const ciphertext = Convert.base64Url(ciphertextB64U).toUint8Array();
+  const authenticationTag = Convert.base64Url(
+    authenticationTagB64U
+  ).toUint8Array();
+
+  // The cipher expects the encrypted data and tag to be concatenated.
+  const ciphertextAndTag = new Uint8Array([
+    ...ciphertext,
+    ...authenticationTag,
+  ]);
+
+  // decrypt using the sharedKey
+  const chacha = xchacha20poly1305(sharedKey, nonce, AAD);
+  const decryptedJwtBytes = chacha.decrypt(ciphertextAndTag);
+  const jwt = Convert.uint8Array(decryptedJwtBytes).toString();
+
+  return jwt;
+}
+
+async function deriveSharedKey(
+  privateKeyDid: BearerDid,
+  publicKeyDid: DidDocument
+) {
+  const exportedClientDid = await privateKeyDid.export();
```

Comment:
`exportedClientDid` is misleading in this generalized function as it is used by both sender and receiver to generate shared key in reverse manner. ie. in the case of `submitAuthResponse()`, the `privateKeyDid` represents the delegated DID, not the ephemeral client DID.

Consider rename to something generic like `privatePortableDid`.

Reply:
good catch. changed